src/dotty/tools/dotc/transform/ClassOf.scala

```diff
@@ -0,0 +1,44 @@
+package dotty.tools.dotc
+package transform
+
+import dotty.tools.dotc.ast.tpd
+import dotty.tools.dotc.core.Constants.Constant
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.core.Decorators.StringDecorator
+import dotty.tools.dotc.core.Symbols._
+import dotty.tools.dotc.core.{TypeErasure, _}
+import dotty.tools.dotc.transform.TreeTransforms.{MiniPhaseTransform, TransformerInfo}
+
+class ClassOf extends MiniPhaseTransform {
+  import tpd._
+
+  override def phaseName: String = "classOf"
+
+  override def transformTypeApply(tree: TypeApply)(implicit ctx: Context, info: TransformerInfo): Tree = {
+    val predefModule = ctx.definitions.ScalaPredefModule
+    val classOfMethod = ctx.requiredMethod(predefModule.moduleClass.asClass, "classOf".toTermName)
+
+    if (tree.symbol eq classOfMethod) {
+      val tp = tree.args.head.tpe
+      val defn = ctx.definitions
+      val claz = tp.classSymbol
+
+      if (ValueClasses.isDerivedValueClass(claz)) {
+        Literal(Constant(ref(claz.companionModule).tpe))
```

Comment:
this is a wrong class you are giving. It should be the class of the value class itself, not it's companion object.